python/usb_console/radio_session.py

```diff
@@ -59,27 +72,78 @@ def check_for_downlink(self):
-            pass
+            #.search() searches from mail. Data gives id's of all emails.
+            type, data = mail.search(None, '(FROM "sbdservice@sbd.iridium.com")', '(UNSEEN)')
+            mail_ids = data[0]
+            id_list = mail_ids.split()
+
+            for num in id_list:
+                #.fetch() fetches the mail for given id where 'RFC822' is an Internet 
+                # Message Access Protocol.
+                typ, data = mail.fetch(num,'(RFC822)')
+                
+                #go through each component of data and print sender and subject line
+                for response_part in data:
+                    if isinstance(response_part, tuple):
+                        # converts message from byte literal to string removing b''
+                        msg = email.message_from_string(response_part[1].decode('utf-8'))
+                        #get the file attached to the message
+                        for part in msg.walk():
+                            
+                            if part.get_content_maintype() == 'multipart':
+                                continue
+
+                            if part.get('Content-Disposition') is None:
+                                continue
+                            
+                            #check if there is an attachment
+                            if part.get_filename() is not None:
+                                #get json data from email attachment
+                                attachmentContents=part.get_payload(decode=True).decode('utf8')
+                                data=json.loads(attachmentContents)
+                                
+                                #iterate through each statefield key of the dictionary
+                                logline=""
+                                for key in self.statefields:
+                                    #if the value for a statefield exists in the json object recieved, update the dictionary and print it to the logger
+                                    if data.get(key)is not None:
+                                        newvalue=data[key]
+                                        self.statefields[key]=newvalue
+                                        logline+=key+": "+str(newvalue)+"\n"
+                                self.logger.put(logline)
 
     def read_state(self, field, timeout=None):
         '''
         Read state.
         
         Read the latest value of the state field received via downlink with the name "field".
         '''
-        raise NotImplementedError
+        return self.statefields[field]
 
     def write_multiple_states(self, fields, vals, timeout=None):
         '''
         Uplink multiple state variables. Return success of write.
         '''
         assert len(fields) == len(vals)
 
-        # TODO send uplink via email.
-        raise NotImplementedError
+        #update statefield dictionary
+        for i in range(len(fields)):
```

Comment:
This is actually not what we want to do. We're not going to be sending all of the state fields up to the satellite; we'll only send the ones specified in the arguments of this function. I would remove this for loop and instead construct a new dictionary based only on `fields` and `vals`.